File: java/sdk/sdk-core/src/main/java/com/novax/sdk/core/http/SdkRequest.java
package com.novax.sdk.core.http;

import com.novax.sdk.core.json.JsonMapper;
import com.novax.sdk.core.request.ApiRequest;
import com.novax.sdk.core.request.HttpMethod;

import java.net.URI;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Wire-level request — what the transport layer actually sends. Built from an
 * {@link ApiRequest} via {@link #from(URI, ApiRequest, JsonMapper)}; immutable
 * thereafter.
 *
 * <p>{@code headers} is {@code Map<String, List<String>>} because HTTP allows
 * the same header name to appear multiple times (RFC 7230). Use
 * {@link #appendHeaders(Map)} to add additional values (multi-value semantics)
 * or {@link #setHeaders(Map)} to replace any existing values for the given keys.
 */
public record SdkRequest(
        HttpMethod method,
        URI uri,
        Map<String, List<String>> headers,
        byte[] body
) {

    public static SdkRequest from(URI endpoint, ApiRequest<?> req, JsonMapper jsonMapper) {
        URI uri = buildUri(endpoint, req.path(), req.queryParams());
        LinkedHashMap<String, List<String>> headers = new LinkedHashMap<>();
        for (var e : req.headers().entrySet()) {
            headers.computeIfAbsent(e.getKey(), k -> new ArrayList<>()).add(e.getValue());
        }
        byte[] body = null;
        if (req.body() != null) {
            body = jsonMapper.writeBytes(req.body());
            headers.computeIfAbsent("Content-Type", k -> new ArrayList<>(List.of("application/json")));
        }
        return new SdkRequest(req.method(), uri, freeze(headers), body);
    }

    /**
     * Append each value to the list for its key (HTTP-correct multi-value).
     */
    public SdkRequest appendHeaders(Map<String, String> additional) {
        LinkedHashMap<String, List<String>> merged = mutableCopy(headers);
        for (var e : additional.entrySet()) {
            merged.computeIfAbsent(e.getKey(), k -> new ArrayList<>()).add(e.getValue());
        }
        return new SdkRequest(method, uri, freeze(merged), body);
    }

    /**
     * Replace any existing values for these keys (single-valued semantics).
     */
    public SdkRequest setHeaders(Map<String, String> overrides) {
        LinkedHashMap<String, List<String>> merged = mutableCopy(headers);
        for (var e : overrides.entrySet()) {
            merged.put(e.getKey(), new ArrayList<>(List.of(e.getValue())));
        }
        return new SdkRequest(method, uri, freeze(merged), body);
    }

    private static LinkedHashMap<String, List<String>> mutableCopy(Map<String, List<String>> src) {
        LinkedHashMap<String, List<String>> copy = new LinkedHashMap<>();
        for (var e : src.entrySet()) {
            copy.put(e.getKey(), new ArrayList<>(e.getValue()));
        }
        return copy;
    }

    private static Map<String, List<String>> freeze(Map<String, List<String>> src) {
        LinkedHashMap<String, List<String>> copy = new LinkedHashMap<>();
        for (var e : src.entrySet()) {
            copy.put(e.getKey(), List.copyOf(e.getValue()));
        }
        return Map.copyOf(copy);
    }

    private static URI buildUri(URI base, String path, Map<String, ?> query) {
        StringBuilder sb = new StringBuilder(base.toString());
        if (sb.charAt(sb.length() - 1) == '/' && path.startsWith("/")) {
            sb.setLength(sb.length() - 1);
        }
        sb.append(path);
        if (query != null && !query.isEmpty()) {
            String qs = query.entrySet().stream()
                    .filter(e -> e.getValue() != null)
                    .map(e -> encode(e.getKey()) + "=" + encode(String.valueOf(e.getValue())))
                    .collect(Collectors.joining("&"));
            if (!qs.isEmpty()) {
                sb.append('?').append(qs);
            }
        }
        return URI.create(sb.toString());
    }

    private static String encode(String s) {
        return URLEncoder.encode(s, StandardCharsets.UTF_8);
    }
}

File: python/novax_sdk/http/sdk_request.py
from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Optional
from urllib.parse import urlencode, quote

from novax_sdk.request.api_request import ApiRequest, HttpMethod


@dataclass
class SdkRequest:
    """Wire-level request. headers is dict[str, list[str]] — HTTP allows the
    same name multiple times (RFC 7230). Use append_headers for multi-value
    semantics and set_headers for single-value replacement."""

    method: HttpMethod
    url: str
    headers: dict[str, list[str]]
    body: Optional[bytes]

    @staticmethod
    def from_api_request(endpoint: str, req: ApiRequest) -> "SdkRequest":
        url = _build_url(endpoint, req.path, req.query_params())
        headers: dict[str, list[str]] = {}
        for k, v in req.headers().items():
            headers.setdefault(k, []).append(v)
        body = None
        if req.body() is not None:
            body = json.dumps(req.body(), ensure_ascii=False).encode()
            headers.setdefault("Content-Type", []).append("application/json")
        return SdkRequest(method=req.method, url=url, headers=headers, body=body)

    def append_headers(self, extra: dict[str, str]) -> "SdkRequest":
        merged = {k: list(v) for k, v in self.headers.items()}
        for k, v in extra.items():
            merged.setdefault(k, []).append(v)
        return SdkRequest(self.method, self.url, merged, self.body)

    def set_headers(self, overrides: dict[str, str]) -> "SdkRequest":
        merged = {k: list(v) for k, v in self.headers.items()}
        for k, v in overrides.items():
            merged[k] = [v]
        return SdkRequest(self.method, self.url, merged, self.body)

    def flat_headers(self) -> list[tuple[str, str]]:
        """Flatten to (name, value) pairs for requests library."""
        return [(k, v) for k, vs in self.headers.items() for v in vs]


def _build_url(endpoint: str, path: str, query: dict[str, Any]) -> str:
    base = endpoint.rstrip("/")
    if not path.startswith("/"):
        path = "/" + path
    url = base + path
    params = {k: v for k, v in query.items() if v is not None}
    if params:
        url += "?" + urlencode(params)
    return url
